### PR TITLE
implement --tree -d

### PR DESCRIPTION
**Pull request by 0jdxt**
Was looking at #295 and decided to have a crack

### CODE REVIEW

**Comment by 0jdxt** [comment]
there is probably a cleaner way to do it and have the logic together, ill have a look

**Comment by 0jdxt** [comment]
I'm not sure why I get the clippy errors, I dont get them locally

**Comment by meain** [comment]
Hey, there has been a relatively big change in how flags are managed due to the introduction of config file. Could you rebase this on top of current master.

**Comment by 0jdxt** [comment]
i can only reproduce the clippy error on 1.43.1, not 1.46

**Comment by 0jdxt** [comment]
@meain this is ready if you're happy with it

**Comment by zwpaper** [comment]
maybe we should do a rebase or squash to remove the unnecessary commits before merging this?

it would be more complicated for others to rebase on top of many small commit and got an unclear commit history

**Comment by 0jdxt** [comment]
Yeah just squash it

File: src/core.rs
Comment (meain) [inline src/core.rs:135]: Something like this would be better:
```rust
matches!(self.flags.layout, Layout::Tree {..})
```

**Review by meain — CHANGES_REQUESTED** [review]
Hey, I think there is a small issue with the implementation.
If we have `tree` as layout in the config but pass `-d` flag as cli arg, it is not getting treated as `--tree -d`.

I think it would be better to not touch how `Layout::Tree` or `dir_only` is parsed, but just do the check for if `dir_only` is set when displaying in tree mode. As in no need to add `dir_only` to `Layout::Tree`.

**Comment by 0jdxt** [comment]
done!

File: src/meta/mod.rs
Comment (meain) [inline src/meta/mod.rs:63]: Hey, I just realized that this might cause issues if we have something like below:
```
mkdir one
ln -s one one/two
```

This will go on a endless loop if we have dereference enabled. I am not really certain if we have to handle it though. I took a look at gnu ls and even thought it does not have `--tree` or `-R` it does not seem to handle the case were we symlink two itmes to each other and just error out with os error.
```
ln -s one two
ln -s two one
```

Do you have any other suggestion? If not, I think we can stick to this as of now hoping that users would be aware of `-L` use with `-R` or `--tree`. Also, when I tested it, it seemed to stop at few levels deep. I am not sure if this is an OS level restriction, but at least `lsd` does not go bonkers and eat up all the cpu if this happens.

**Review by meain — COMMENTED** [review]
Could you also add a few more tests, just to avoid any future regressions.

- for --tree in config and -d in cli arg
- for --tree -l -d
- for case were symlink dir (do not show content without `-L`)

File: src/meta/mod.rs
Comment (meain) [reply src/meta/mod.rs:63]: Just realized that gnu ls has `-R`.

```
/tmp/tree-test/one:
two
gls: /tmp/tree-test/one/two: not listing already-listed directory
```

This is the error that we get. Maybe we should do this check.

Comment (0jdxt) [reply src/meta/mod.rs:63]: Yes I think we should take from `ls` in general so ideally people could replace `ls` with `lsd` and behaviour is as expected maybe with some exceptions.

I will look into some way of "remembering" displayed files, and at how `ls` does it. Also, `tree` gives the following which I will try to recreate. Also perhaps we could add a "X directories, Y files" line too for `lsd --tree`

```bash
$ tree
.
└── one
    └── two -> ../one  [recursive, not followed]

2 directories, 0 files
```

Comment (meain) [reply src/meta/mod.rs:63]: Btw, as of now it looks like `--tree` follows the symlink, but `tree` does not seem to follow symlinks by default. I think it would be better for us to do it that way too and only follow symlinks if there is `-L` with tree.

Comment (meain) [reply src/meta/mod.rs:63]: "recursive" might not be the proper term. Like gnu ls, something like "already-listed" would be better as we probably need to make it so that all previously visited items are not listed, not just recursive (tracing if something is recursive might be unnecessary work).

Comment (meain) [reply src/meta/mod.rs:63]: Hey, the changes to incorporate "remembering" visited paths, I guess, can be a separate  PR if that works for you. I can merge this in for now if you could add the rest of the tests as well.

**Comment by meain** [comment]
Closing in favor of https://github.com/Peltoche/lsd/pull/456 . Thanks @0jdxt for getting it started.

**Pull request closed without merging**
